File: gateway/src/metrics.js
// Minimal Prometheus text-format metrics.
const esc = (v) => String(v).replace(/\\/g, '\\\\').replace(/"/g, '\\"').replace(/\n/g, '\\n');
const fmt = (labels) => {
  const entries = Object.entries(labels);
  return entries.length ? `{${entries.map(([k, v]) => `${k}="${esc(v)}"`).join(',')}}` : '';
};

export class Metrics {
  constructor() {
    this.counters = new Map();
    this.gauges = new Map();
    this.histograms = new Map();
    this.help = new Map();
  }

  counter(name, help) { this.help.set(name, ['counter', help]); if (!this.counters.has(name)) this.counters.set(name, new Map()); }
  gauge(name, help, fn) { this.help.set(name, ['gauge', help]); this.gauges.set(name, fn); }
  histogram(name, help, buckets = [0.1, 0.5, 1, 2.5, 5, 10, 30, 60, 120]) {
    this.help.set(name, ['histogram', help]);
    this.histograms.set(name, { buckets, series: new Map() });
  }

  inc(name, labels = {}, value = 1) {
    const series = this.counters.get(name);
    const key = JSON.stringify(labels);
    series.set(key, (series.get(key) || 0) + value);
  }

  observe(name, value, labels = {}) {
    const h = this.histograms.get(name);
    const key = JSON.stringify(labels);
    const s = h.series.get(key) || { counts: h.buckets.map(() => 0), sum: 0, count: 0 };
    h.buckets.forEach((b, i) => { if (value <= b) s.counts[i] += 1; });
    s.sum += value;
    s.count += 1;
    h.series.set(key, s);
  }

  value(name, labels = {}) {
    return this.counters.get(name)?.get(JSON.stringify(labels)) || 0;
  }

  render() {
    const out = [];
    for (const [name, [type, help]] of this.help) {
      out.push(`# HELP ${name} ${help}`, `# TYPE ${name} ${type}`);
      if (type === 'counter') for (const [k, v] of this.counters.get(name)) out.push(`${name}${fmt(JSON.parse(k))} ${v}`);
      if (type === 'gauge') out.push(`${name} ${Number(this.gauges.get(name)()) || 0}`);
      if (type === 'histogram') {
        const h = this.histograms.get(name);
        for (const [k, s] of h.series) {
          const labels = JSON.parse(k);
          h.buckets.forEach((b, i) => out.push(`${name}_bucket${fmt({ ...labels, le: b })} ${s.counts[i]}`));
          out.push(`${name}_bucket${fmt({ ...labels, le: '+Inf' })} ${s.count}`, `${name}_sum${fmt(labels)} ${s.sum}`, `${name}_count${fmt(labels)} ${s.count}`);
        }
      }
    }
    return out.join('\n') + '\n';
  }
}

File: bee2bee/metrics.py
"""Minimal Prometheus text-format metrics (no external dependency)."""

from __future__ import annotations

import threading
from typing import Dict, Iterable, List, Optional, Tuple

LabelKey = Tuple[Tuple[str, str], ...]


def _key(labels: Optional[Dict[str, str]]) -> LabelKey:
    return tuple(sorted((labels or {}).items()))


def _fmt_labels(key: LabelKey, extra: Iterable[Tuple[str, str]] = ()) -> str:
    items = list(key) + list(extra)
    if not items:
        return ""
    body = ",".join(f'{k}="{str(v).replace(chr(92), chr(92) * 2).replace(chr(34), chr(92) + chr(34))}"' for k, v in items)
    return "{" + body + "}"


class _Metric:
    kind = "untyped"

    def __init__(self, name: str, help_text: str):
        self.name = name
        self.help = help_text
        self._lock = threading.Lock()

    def render(self) -> List[str]:
        return [f"# HELP {self.name} {self.help}", f"# TYPE {self.name} {self.kind}"] + self._samples()

    def _samples(self) -> List[str]:
        raise NotImplementedError


class Counter(_Metric):
    kind = "counter"

    def __init__(self, name: str, help_text: str):
        super().__init__(name, help_text)
        self._values: Dict[LabelKey, float] = {}

    def inc(self, amount: float = 1.0, **labels: str) -> None:
        k = _key(labels)
        with self._lock:
            self._values[k] = self._values.get(k, 0.0) + amount

    def value(self, **labels: str) -> float:
        return self._values.get(_key(labels), 0.0)

    def _samples(self) -> List[str]:
        return [f"{self.name}{_fmt_labels(k)} {v}" for k, v in sorted(self._values.items())]


class Gauge(_Metric):
    kind = "gauge"

    def __init__(self, name: str, help_text: str, fn=None):
        super().__init__(name, help_text)
        self._values: Dict[LabelKey, float] = {}
        self._fn = fn

    def set(self, value: float, **labels: str) -> None:
        with self._lock:
            self._values[_key(labels)] = value

    def _samples(self) -> List[str]:
        if self._fn is not None:
            return [f"{self.name} {float(self._fn())}"]
        return [f"{self.name}{_fmt_labels(k)} {v}" for k, v in sorted(self._values.items())]


class Histogram(_Metric):
    kind = "histogram"
    DEFAULT_BUCKETS = (0.05, 0.1, 0.25, 0.5, 1, 2.5, 5, 10, 30, 60, 120, 300)

    def __init__(self, name: str, help_text: str, buckets: Iterable[float] = DEFAULT_BUCKETS):
        super().__init__(name, help_text)
        self.buckets = tuple(sorted(buckets))
        self._data: Dict[LabelKey, Tuple[List[int], float, int]] = {}

    def observe(self, value: float, **labels: str) -> None:
        k = _key(labels)
        with self._lock:
            counts, total, n = self._data.get(k, ([0] * len(self.buckets), 0.0, 0))
            for i, b in enumerate(self.buckets):
                if value <= b:
                    counts[i] += 1
            self._data[k] = (counts, total + value, n + 1)

    def _samples(self) -> List[str]:
        out: List[str] = []
        for k, (counts, total, n) in sorted(self._data.items()):
            for b, c in zip(self.buckets, counts, strict=True):
                out.append(f"{self.name}_bucket{_fmt_labels(k, [('le', str(b))])} {c}")
            out.append(f"{self.name}_bucket{_fmt_labels(k, [('le', '+Inf')])} {n}")
            out.append(f"{self.name}_sum{_fmt_labels(k)} {total}")
            out.append(f"{self.name}_count{_fmt_labels(k)} {n}")
        return out


class Registry:
    def __init__(self) -> None:
        self._metrics: List[_Metric] = []

    def register(self, metric: _Metric) -> _Metric:
        self._metrics.append(metric)
        return metric

    def counter(self, name: str, help_text: str) -> Counter:
        return self.register(Counter(name, help_text))  # type: ignore[return-value]

    def gauge(self, name: str, help_text: str, fn=None) -> Gauge:
        return self.register(Gauge(name, help_text, fn))  # type: ignore[return-value]

    def histogram(self, name: str, help_text: str, buckets: Iterable[float] = Histogram.DEFAULT_BUCKETS) -> Histogram:
        return self.register(Histogram(name, help_text, buckets))  # type: ignore[return-value]

    def render(self) -> str:
        lines: List[str] = []
        for metric in self._metrics:
            lines.extend(metric.render())
        return "\n".join(lines) + "\n"
